proposals/simd/NewOpcodes.md

```diff
@@ -0,0 +1,124 @@
+| i8x16.shr_u          | 0x6d   | i16x8.shr_u              | 0x8d   | i32x4.shr_u              | 0xad   | i64x2.shr_u | 0xcd   |
+| i8x16.add            | 0x6e   | i16x8.add                | 0x8e   | i32x4.add                | 0xae   | i64x2.add   | 0xce   |
+| i8x16.add_saturate_s | 0x6f   | i16x8.add_saturate_s     | 0x8f   | ---- add_sat ----        | 0xaf   | ----        | 0xcf   |
+| i8x16.add_saturate_u | 0x70   | i16x8.add_saturate_u     | 0x90   | ---- add_sat ----        | 0xb0   | ----        | 0xd0   |
+| i8x16.sub            | 0x71   | i16x8.sub                | 0x91   | i32x4.sub                | 0xb1   | i64x2.sub   | 0xd1   |
+| i8x16.sub_saturate_s | 0x72   | i16x8.sub_saturate_s     | 0x92   | ---- sub_sat ----        | 0xb2   | ----        | 0xd2   |
+| i8x16.sub_saturate_u | 0x73   | i16x8.sub_saturate_u     | 0x93   | ---- sub_sat ----        | 0xb3   | ----        | 0xd3   |
+| ---- dot ----        | 0x74   | ---- dot ----            | 0x94   | i32x4.dot_i16x8_s        | 0xb4   | ----        | 0xd4   |
+| ---- mul ----        | 0x75   | i16x8.mul                | 0x95   | i32x4.mul                | 0xb5   | i64x2.mul   | 0xd5   |
+| i8x16.min_s          | 0x76   | i16x8.min_s              | 0x96   | i32x4.min_s              | 0xb6   | ----        | 0xd6   |
+| i8x16.min_u          | 0x77   | i16x8.min_u              | 0x97   | i32x4.min_u              | 0xb7   | ----        | 0xd7   |
+| i8x16.max_s          | 0x78   | i16x8.max_s              | 0x98   | i32x4.max_s              | 0xb8   | ----        | 0xd8   |
+| i8x16.max_u          | 0x79   | i16x8.max_u              | 0x99   | i32x4.max_u              | 0xb9   | ----        | 0xd9   |
+| ---- avgr_s ----     | 0x7a   | ---- avgr_s ----         | 0x9a   | ---- avgr_s ----         | 0xba   | ----        | 0xda   |
+| i8x16.avgr_u         | 0x7b   | i16x8.avgr_u             | 0x9b   | ---- avgr_u ----         | 0xbb   | ----        | 0xdb   |
+
+| f32x4 Op        | opcode | f64x2 Op        | opcode |
+| --------------- | ------ | --------------- | ------ |
+| f32x4.abs       | 0xe0   | f64x2.abs       | 0xf0   |
+| f32x4.neg       | 0xe1   | f64x2.neg       | 0xf1   |
+| ---- round ---- | 0xe2   | ---- round ---- | 0xf2   |
+| f32x4.sqrt      | 0xe3   | f64x2.sqrt      | 0xf3   |
+| f32x4.add       | 0xe4   | f64x2.add       | 0xf4   |
+| f32x4.sub       | 0xe5   | f64x2.sub       | 0xf5   |
+| f32x4.mul       | 0xe6   | f64x2.mul       | 0xf6   |
+| f32x4.div       | 0xe7   | f64x2.div       | 0xf7   |
+| f32x4.min       | 0xe8   | f64x2.min       | 0xf8   |
+| f32x4.max       | 0xe9   | f64x2.max       | 0xf9   |
+| ---- pmin ----  | 0xea   | ---- pmin ----  | 0xfa   |
+| ---- pmax ----  | 0xeb   | ---- pmax ----  | 0xfb   |
+
+| Conversion Op           | opcode   |
+| ----------------------- | -------- |
+| i32x4.trunc_sat_f32x4_s | 0x0100   |
+| i32x4.trunc_sat_f32x4_u | 0x0101   |
+| f32x4.convert_i32x4_s   | 0x0102   |
+| f32x4.convert_i32x4_u   | 0x0103   |
```

Comment:
Is there any way to put this section somewhere else to avoid double-byte encoding? It feels like it would make encoding more uniform, would make opcode-based lookup tables have a small fixed size without extra range checks, and would make it possible to patch the bytecode in place vs the old encoding for compatibility purposes.

Reply:
Ohh, or is it the case that right now the opcode is actually not encoded as a byte, since LEB128 encoding uses two bytes for any opcodes >=0x80? :-/

Reply:
Yes, all the bytecodes are LEB128 so we already have opcodes that take 2 bytes in the current numbering.

Reply:
This is unfortunate. The existing Wasm spec doesn't do this as far as I understand. We have 170 opcodes right now. I'm not sure why the LEB128 encoding was chosen, is there a precedent for this elsewhere in Wasm?

Reply:
I had the same question in #46. TL;DR: this is how all proposals work since 2017, but SIMD is the only one where you can tell the difference. Notably, V8 was mistakenly using bytes instead of LEB128s until a couple weeks ago: https://bugs.chromium.org/p/v8/issues/detail?id=10258.

Reply:
So to clarify, the baseline wasm spec doesn't use LEB128 for instructions but all extensions do? A bit confused as to how v8 managed to work with byte encoding if there are a lot of opcodes today over 0x80?

Reply:
Also I'm a bit sad that we're doing LEB128. It hurts code size (maybe in a minor way), it makes decoding slightly more involved (why?).

If we introduce *more* opcodes, we need to be introducing them in groups as we discussed in prior calls, because just adding new opcodes incrementally is impossible to deal with from the application side.

So on the encoding perspective, it seems straightforward to dedicate, say, 0xf0-0xff to prefixes for future extensions. This can work well in that when we introduce a fast SIMD subset, it's just going to take 0xf1 or smth and evolve that independently from any other subsets.

Of course there's a risk of prefix-based approach eventually evolving into something odd, and the parsing of extensions gets more uniform if we just go with LEB128. Still, it's odd to see this diverging from the Wasm spec approach.

Reply:
I'm still trying to understand this (sorry for the spam). You're saying that the spec says that opcodes should be encoded with LEB128 but the *entire ecosystem* chose to use bytes right now? That is - this isn't just a v8 issue, it's also an issue on LLVM side, wabt, et al - *all* tools that interact with SIMD must be assuming bytes?

Reply:
Dear god. Thanks, this makes more sense now!

Reply:
> I had the same question in #46. TL;DR: this is how all proposals work since 2017, but SIMD is the only one where you can tell the difference. Notably, V8 was mistakenly using bytes instead of LEB128s until a couple weeks ago: https://bugs.chromium.org/p/v8/issues/detail?id=10258.

Same bug in Firefox, found it when I implemented SIMD :-}

Reply:
Looking at this again as I'm updating engine opcodes, I would also like to express a preference for the decoded opcodes to be 1-byte because we have the space. The reason is that on the engine side, this will add some bloat to opcode/signature tables which assume 1-byte opcodes. It's not terrible to update, but considering we have the space, and it's just the 4 opcodes that are spilling over I'd like to avoid it as it looks like it's possible in this case.